File: app/leetcode/No26.removeDuplicates.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var removeDuplicates = function (nums) {
    // 方法1：利用set去重
    // const set = new Set();
    // let len = 0;
    // for (const num of nums) {
    //     if (!set.has(num)) {
    //         nums[len++] = num;
    //         set.add(num);
    //     }
    // }
    // return len;

    // 方法2，单指针维护当前有效元素的个数
    const n = nums.length;
    if (n === 0) return 0;
    let index = 1;
    for (let i = 1; i < n; i++) {
        const num = nums[i];
        if (num !== nums[i - 1]) {
            nums[index++] = num;
        }
    }
    return index;
};

nums = [1, 1, 2]
nums = [0, 0, 1, 1, 1, 2, 2, 3, 3, 4]
ret = removeDuplicates(nums);
debugger;

File: app/leetcode/No26.removeDuplicates.go
package main

import "fmt"

func removeDuplicates(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	// 利用 set 去重
	set := make(map[int]bool)
	l := 0
	for i := 0; i < len(nums); i++ {
		if !set[nums[i]] {
			set[nums[i]] = true
			nums[l] = nums[i]
			l++
		}
	}
	return l
}

func removeDuplicates2(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	// 题目说是非严格单调递增，因此后一个元素要么一定大于等于前一个元素 即 1,1,2 是可能得， 1,2,2,1 是不可能的
	// 也就是说如果出现重复元素，则重复的元素一定是相邻的
	k := 1
	for i := 1; i < len(nums); i++ {
		if nums[i] != nums[i-1] {
			nums[k] = nums[i]
			k++
		}
	}
	return k
}

func main() {
	nums := []int{0, 0, 1, 1, 1, 2, 2, 3, 3, 4}
	r := removeDuplicates2(nums)
	fmt.Println(r, nums)
	nums = []int{1, 1, 2}
	r = removeDuplicates2(nums)
	fmt.Println(r, nums)
}
